behaviour/graql/language/define.feature

```diff
@@ -230,157 +271,79 @@ Feature: Graql Define Query
-      define employment sub relation, relates employer;
-      """
-    Given the integrity is validated
-
+  Scenario: define 'key' is idempotent
     Given graql define
       """
       define
-      child sub person;
-      person sub entity, plays employer;
-      """
-    Given the integrity is validated
-
-    When get answers of graql query
-      """
-      match $x type child, plays $r; get;
-      """
-    Then concept identifiers are
-      |             | check | value            |
-      | EMPLOYEE    | label | employee         |
-      | EMPLOYER    | label | employer         |
-      | EARNER      | label | earner           |
-      | NAME_OWNER  | label | @has-name-owner  |
-      | EMAIL_OWNER | label | @key-email-owner |
-      | CHILD       | label | child            |
-    Then uniquely identify answer concepts
-      | x     | r           |
-      | CHILD | EMPLOYEE    |
-      | CHILD | EMPLOYER    |
-      | CHILD | EARNER      |
-      | CHILD | NAME_OWNER  |
-      | CHILD | EMAIL_OWNER |
-
-
-  @ignore
-  # re-enable when we can query schema 'has' and 'key' with variables eg: 'match $x type ___, has key $a; get;'
-  Scenario: define additional 'has' is visible from all children
-    Given graql define
-    """
-       define
-       child sub person;
-       phone-number sub attribute, value long;
-       person sub entity, has phone-number;
+      address sub attribute, value string;
+      house sub entity, key address, key address, key address;
       """
     Given the integrity is validated
-
     When get answers of graql query
       """
-      match $x type child, has $y; get;
+      match $x key address; get;
       """
     Then concept identifiers are
-      |       | check | value        |
-      | CHILD | label | child        |
-      | NAME  | label | name         |
-      | PHONE | label | phone-number |
+      |     | check | value |
+      | HOU | label | house |
     Then uniquely identify answer concepts
-      | x     | y     |
-      | CHILD | NAME  |
-      | CHILD | PHONE |
-
+      | x   |
+      | HOU |
 
-  @ignore
-  # re-enable when we can query schema 'has' and 'key' with variables eg: 'match $x type ___, has key $a; get;'
-  Scenario: define additional 'key' is visible from all children
-    Given graql define
-      """
-      define
-      child sub person;
-      phone-number sub attribute, value long;
-      person sub entity, key phone-number;
-      """
-    Given the integrity is validated
 
-    When get answers of graql query
+  Scenario: define concept without 'sub' throws
+    Then graql define throws
       """
-      match $x type child, key $y; get;
+      define flying-spaghetti-monster;
       """
-    Then concept identifiers are
-      |       | check | value |
-      | CHILD | label | child |
-      | EMAIL | label | email |
-    Then uniquely identify answer concepts
-      | x     | y     |
-      | CHILD | EMAIL |
-      | CHILD | EMAIL |
 
 
   #############
```

Comment:
I feel we're missing a section on attributes, including the simple stuff like being able to define an attribute of each type, and that each type of attribute can be owned. Then perhaps there are some more fundamental attribute cases I can't think of yet. This may be overkill, I'll leave it to you, but attributes often present difficulties, and with the multiple datatypes possible they, in some ways, present more opportunities for failure than entities and relations.

Reply:
I think you're right that some of the basic stuff should be tested individually against each attribute type.

I had the same decisions to make when it came to test duplication across entity/relation/attribute. "Is this test significant enough to justify rewriting it for each individual kind of type?" I duplicated the really fundamental cases, like "a type can be marked as `abstract`", and "a type inherits `plays` from its parent"; and of course, the all-important "defining a (entity|relation|attribute) creates a type". So it probably makes sense to run through the very basics of each attribute value type as well.

Reply:
Added:

```
an attribute with type `string` can be defined
an attribute with type `long` can be defined
an attribute with type `double` can be defined
an attribute with type `boolean` can be defined
an attribute with type `datetime` can be defined
a type can `has` an attribute of type `string`
a type can `has` an attribute of type `long`
a type can `has` an attribute of type `double`
a type can `has` an attribute of type `boolean`
a type can `has` an attribute of type `datetime`
```